Sprint 10

dreamWillAppear opened the pull request:
ПР для сдачи задания по 10 спринту

dreamWillAppear committed:
added constants with access keys

dreamWillAppear committed:
created WebViewViewController for further creation of authorization into app

dreamWillAppear committed:
add a login screen with UIProgressView

dreamWillAppear committed:
test

Commit:
implemented authorization and storage of authorization token in UserDefaults

dreamWillAppear committed:
Added an Unsplash screen to check for authorization.
Added authorization functionality with OAuth2.

Commit:
did other improvements

helecta commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,44 @@
+    private let webViewController = WebViewViewController()
+    private let oAuth2Service = OAuth2Service.shared
+    private let oAuth2TokenStorage = OAuth2TokenStorage.shared
+    weak var delegate: AuthViewControllerDelegateProtocol?
```
Не забывай организовывать код следующим образом:

- Публичные переменные
- IBOutlet
- Приватные переменные
- Публичные методы(сначала `init`ы, `override`ы, и потом остальные)
- Приватные методы
- IBAction

Подробнее почитать можно тут: [Организация кода](https://github.com/RedMadRobot/RMR-swift-style-guide#code-organization)

helecta commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,44 @@
+        oAuth2Service.fetchOAuthToken(code: code) { result in
+            switch result {
+            case .success(let token):
+                self.oAuth2TokenStorage.token = token
```
В замыкание для избежания retain cycl'ов, необходимо передавать слабую ссылку на `self`

helecta commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,144 @@
+        String?,
+        of object: Any?,
+        change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
+            if keyPath == #keyPath(WKWebView.estimatedProgress) {
```
Не забывай следить за отступами, общепринятые отступы помогают читать код быстрее. В общем случае поможет выделить весь файл cmd+A и зажать control+i.

helecta commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,144 @@
+        }
+    
+    private func updateProgress() {
+        progressView.progress = Float(webView.estimatedProgress)
```
если использовать progressView.setProgress(…), то можно сделать заполнение прогресс бара более плавным.

helecta commented on ImageFeed/Protocols+Models/OAuthTokenResponseBodyModel.swift:
```diff
@@ -0,0 +1,26 @@
+    let scope: String
+    let createdAt: Int
+    
+    enum CodingKeys: String, CodingKey {
```
Еще можно таким образом настроить JSONDecoder: https://nemecek.be/blog/90/how-to-decode-snake-case-with-codable  
Тогда CodingKeys будет не нужно

helecta commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,144 @@
+    }
+    
+    private func configureProgressView() {
+        progressView.progressViewStyle = .bar
```
progressView.progressViewStyle должен быть .default. Тогда будет как в макете

dreamWillAppear committed:
- did code reorganization;
- added weak reference to self in closure;
- func updateProgress() reconfigured with .setProgress(...);
- JSONDecoder  reconfigured with .convertFromSnakeCase;
- reconfigured  progressView.progressViewStyle  from .bar to .default

dreamWillAppear commented:
Привет! 
Большое спасибо за комментарии. Поправил, ПР выслал.

dreamWillAppear committed:
some layout edits
